// postgres-transaction/transaction.js
module.exports = (function() {
	function transaction(connection, done, callback) {
		connection.query('BEGIN', didBeginTransaction);
		function didBeginTransaction(err) {
			if(err) {
				return rollback(rollbackCallback(err, callback));
			}
			function rollback(rollbackCallback) {
				connection.query('ROLLBACK', didRollback);
				function didRollback(err) {
					if(err) {
						throw err;
					}
					done(err);
					return rollbackCallback(err); 
				}
			}
		
			function query(preparedStatement, callback) {
				connection.query(preparedStatement, performedQuery);
				function performedQuery(err, result) {
					if(err) {
						return rollback(rollbackCallback(err, callback));				
					}
					callback(err, result);
				}
			}
			
			function commit(callback) {
				connection.query('COMMIT', didCommit);
				function didCommit(err, result) {
					if(callback) {
						callback(err, result);
					}
					done(err, result);
				}
			}
			
			function rollbackCallback(err, callback) {
				return function(rollbackError) {
					callback(err);
				}
			}
			
			var transaction = {
				query: query,
				commit: commit,
				rollback: rollback
			};
			callback(err, transaction);
		}
	}

	return transaction;
})()
